cpp/src/io/parquet/page_data.cu

```diff
@@ -1610,123 +1623,163 @@ __device__ void gpuDecodeLevels(page_state_s* s,
+      start_depth, end_depth, d, s, rep, def, value_count, value_count + batch_size, t);
 
-    // is this thread within row bounds? on the first pass we don't know the bounds, so we will be
-    // computing the full size of the column.  on the second pass, we will know our actual row
-    // bounds, so the computation will cap sizes properly.
+    // is this thread within row bounds? in the non skip_rows/num_rows case this will always
+    // be true.
     int in_row_bounds = 1;
+
+    // if we are in the skip_rows/num_rows case, we need to check against these limits
     if (bounds_set) {
-      // absolute row index
-      int32_t thread_row_index =
-        input_row_count + ((__popc(warp_row_count_mask & ((1 << t) - 1)) + is_new_row) - 1);
-      in_row_bounds = thread_row_index >= s->row_index_lower_bound &&
-                          thread_row_index < (s->first_row + s->num_rows)
-                        ? 1
-                        : 0;
-
-      uint32_t const row_bounds_mask  = ballot(in_row_bounds);
-      int const first_thread_in_range = __ffs(row_bounds_mask) - 1;
-
-      // if we've found the beginning of the first row, mark down the position
-      // in the def/repetition buffer (skipped_values) and the data buffer (skipped_leaf_values)
-      if (!t && first_thread_in_range >= 0 && s->page.skipped_values < 0) {
-        // how many values we've skipped in the rep/def levels
-        s->page.skipped_values = input_value_count + first_thread_in_range;
-        // how many values we've skipped in the actual data stream
-        s->page.skipped_leaf_values =
-          input_leaf_count + __popc(warp_leaf_count_mask & ((1 << first_thread_in_range) - 1));
+      // get absolute thread row index
+      int const is_new_row = start_depth == 0;
+      int thread_row_count, block_row_count;
+      block_scan(temp_storage.scan_storage)
+        .InclusiveSum(is_new_row, thread_row_count, block_row_count);
+      __syncthreads();
+
+      // get absolute thread leaf index
+      int const is_new_leaf = (d >= s->nesting_info[max_depth - 1].max_def_level);
+      int thread_leaf_count, block_leaf_count;
+      block_scan(temp_storage.scan_storage)
+        .InclusiveSum(is_new_leaf, thread_leaf_count, block_leaf_count);
+      __syncthreads();
+
+      // if this thread is in row bounds
+      int const row_index = (thread_row_count + row_count) - 1;
+      in_row_bounds =
+        (row_index >= s->row_index_lower_bound) && (row_index < (s->first_row + s->num_rows));
+
+      // if we have not set skipped values yet, see if we found the first in-bounds row
+      if (!skipped_values_set) {
+        int local_count, global_count;
+        block_scan(temp_storage.scan_storage)
+          .InclusiveSum(in_row_bounds, local_count, global_count);
+        __syncthreads();
+
+        // we found it
+        if (global_count > 0) {
+          // this is the thread that represents the first row.
+          if (local_count == 1) {
```

Comment:
```suggestion
          if (local_count == 1 && in_row_bounds) {
```
I had an issue in testing with setting `num_rows` to `1`.  In that case, `local_count` will equal `1` for every thread above the target thread, leading to a race when setting `skipped_values`.  Testing for `in_row_bounds` fixed that.

Reply:
What's the repro case?  Would a column with say, 100 rows, selecting any 1 row out of it do the trick?

Reply:
That nested file I sent you before, try skip_rows=100 and num_rows=1.  Run that a bunch of times and check the returned row.  I get different output every now and then.

Reply:
or for fun, put a print statement inside `if (local_count == 1) {` and see how many threads raise their hands :)

Reply:
Confirmed.  List<int> column with rows of size 1, skip rows 1, num_rows 1.  Thanks for the find!  Saved a lot of headaches down the road.